src/hotspot/share/gc/shenandoah/shenandoahUncommitThread.cpp

```diff
@@ -0,0 +1,147 @@
+
+      // Explicit GC tries to uncommit everything down to min capacity.
+      // Soft max change tries to uncommit everything down to target capacity.
+      // Periodic uncommit tries to uncommit suitable regions down to min capacity.
+      if (has_work(shrink_before, shrink_until)) {
+        uncommit(shrink_before, shrink_until);
+        last_shrink_time = current;
+      }
+    }
+    {
+      MonitorLocker locker(&_lock, Mutex::_no_safepoint_check_flag);
+      if (!_stop_requested.is_set()) {
+        locker.wait((int64_t )shrink_period);
+      }
+    }
+  }
+}
+
+bool ShenandoahUncommitThread::has_work(double shrink_before, size_t shrink_until) const {
+  // Determine if there is work to do. This avoids taking heap lock if there is
+  // no work available, avoids spamming logs with superfluous logging messages,
+  // and minimises the amount of work while locks are taken.
+
+  if (_heap->committed() <= shrink_until) {
+    return false;
+  }
+
+  for (size_t i = 0; i < _heap->num_regions(); i++) {
+    ShenandoahHeapRegion *r = _heap->get_region(i);
+    if (r->is_empty_committed() && (r->empty_time() < shrink_before)) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+void ShenandoahUncommitThread::notify_soft_max_changed() {
+  if (_soft_max_changed.try_set()) {
+    MonitorLocker locker(&_lock, Mutex::_no_safepoint_check_flag);
+    locker.notify_all();
+  }
+}
+
+void ShenandoahUncommitThread::notify_explicit_gc_requested() {
+  if (_explicit_gc_requested.try_set()) {
+    MonitorLocker locker(&_lock, Mutex::_no_safepoint_check_flag);
+    locker.notify_all();
+  }
+}
+
+void ShenandoahUncommitThread::uncommit(double shrink_before, size_t shrink_until) {
+  assert (ShenandoahUncommit, "should be enabled");
+
+  EventMark em("Concurrent uncommit");
+  double start = os::elapsedTime();
+
+  // Application allocates from the beginning of the heap, and GC allocates at
+  // the end of it. It is more efficient to uncommit from the end, so that applications
+  // could enjoy the near committed regions. GC allocations are much less frequent,
+  // and therefore can accept the committing costs.
+  size_t count = 0;
+  for (size_t i = _heap->num_regions(); i > 0; i--) { // care about size_t underflow
+    ShenandoahHeapRegion* r = _heap->get_region(i - 1);
+    if (r->is_empty_committed() && (r->empty_time() < shrink_before)) {
+      SuspendibleThreadSetJoiner sts_joiner;
+      ShenandoahHeapLocker locker(_heap->lock());
+      if (r->is_empty_committed()) {
+        if (_heap->committed() < shrink_until + ShenandoahHeapRegion::region_size_bytes()) {
+          break;
+        }
+
+        r->make_uncommitted();
+        count++;
+      }
+    }
+    SpinPause(); // allow allocators to take the lock
+  }
+
+  if (count > 0) {
+    _heap->notify_heap_changed();
+    double elapsed = os::elapsedTime() - start;
+    log_info(gc)("Uncommitted " SIZE_FORMAT " regions, in %.3fs", count, elapsed);
```

Comment:
Yeah, at least restore the log format and add `gc+start` log as well.

Reply:
This one is still not addressed, unfortunately ^

Reply:
Yes, I spent some time trying to resurrect `ShenandoahConcurrentPhase` for uncommit here, but it really doesn't want to be used outside of a gc cycle. Also, previously it was logging heap _usage_, which isn't quite what we want here (this may actually increase during this phase, which makes it seem as though nothing is being uncommitted).

I've restored the original logging format, but instead of logging heap usage it is now logging heap committed before and after. Here is an excerpt from specjbb2015 with `-Xms5g -Xmx10g`:

```
[2024-11-20T20:02:25.056+0000][97.396s][22293][info][gc,start       ] Concurrent uncommit
[2024-11-20T20:02:25.072+0000][97.412s][22293][info][gc             ] Concurrent uncommit 5424M->5120M(5120M) 15.988ms
[2024-11-20T20:05:17.916+0000][270.255s][22293][info][gc,start       ] Concurrent uncommit
[2024-11-20T20:05:18.169+0000][270.508s][22293][info][gc             ] Concurrent uncommit 10240M->5120M(5120M) 253.048ms
[2024-11-20T20:06:45.329+0000][357.668s][22293][info][gc,start       ] Concurrent uncommit
[2024-11-20T20:06:45.596+0000][357.935s][22293][info][gc             ] Concurrent uncommit 10240M->5120M(5120M) 267.144ms
[2024-11-20T20:06:57.147+0000][369.486s][22293][info][gc,start       ] Concurrent uncommit
[2024-11-20T20:06:57.148+0000][369.487s][22293][info][gc             ] Concurrent uncommit 5456M->5440M(5440M) 1.189ms
```

Reply:
If we are emitting a log line that looks like a properly formatted GC log line, but the numbers there mean something else for `Concurrent uncommit`, we are bound to confuse users and automatic tools. Uncommit should affect `capacity`, this is how we know how deep we have uncommitted. So, I suggest we emit:

```
Concurrent uncommit XXXXM->XXXXM (YYYYM) z.zzzms
```

...where `XXXX` is the heap used at the end of uncommit (note before and after are the same) and YYYY is capacity. This will not expose users to thinking uncommit grows the heap usage, and would give us instantaneous view on heap usage and capacity.

Reply:
Hmm, the numbers are preceded by `Concurrent uncommit`, with that context it's not much of a stretch to think these numbers represent the change in _committed_ memory. The original log message (in which heap usage may increase during uncommit) was _not_ helpful. A message with the same format in which heap usage also appears to _not change at all_ during an uncommit is also perplexing. Are we trying too hard to preserve the original, not useful message? Maybe we just want a new message that plainly says: 

```
Concurrently uncommitted XXXXM in z.zzzms
```
or
```
Concurrent uncommit: time z.zzzms, committed before XXXXM, committed after YYYYM, capacity ZZZZM
```